### ISSUE TITLE
Add Echo to the Stereo Sound

### ISSUE THREAD

**Issue by Miyakawa-99** [kcat/openal-soft#442]
Hi.
I want to add delay between left and right to the generated 3D sound.
The 3D sound was created by "alSource3f(source, AL_POSITION, x,y,z)".
I added an echo effect to it and confirmed that the delay between left and  right occur.
(I used  AL_ECHO_LRDELAY and AL_ECHO_SPREAD.)
But after adding eho, the pan disappeared.
(I used example/alreverb.c as a reference for adding echoes and loading sounds. )
Is there a way to add an echo effect while still maintaining the stereophonic effect?
Or, if I'm not doing it right, I'd like to know.

**Comment by kcat** [comment]
I'm not quite sure what you're attempting to do. With the echo effect, the normal sound still plays once so it'll maintain stereo panning, then it'll echo to one side then the other (the apparent width and delay between the two echoes dependent on the SPREAD and LRDELAY parameters).

A plain old delay between the left and right output for a 3D sound is essentially what HRTF does. It applies an angle-dependent delay to simulate the sound reaching the two ears at different times when wearing headphones (also includes volume and frequency adjustments to simulate the head itself).

If this isn't what you're meaning, can you clarify what you're trying to do?

**Comment by Miyakawa-99** [comment]
Sorry for my lack of explanation.
I'm trying to play 3-D sound with two stereo speakers.
However, if the listener is off-center of the two speakers, the sound the listener receives will have a slight delay on either side.
I want to limit that effect.
Moreover, I want the delay to change dynamically as the listener moves.
(I have the system which can estimates the listener's position.)
Therefore, I came up with using echo effect to fill the  gap.

In other words, the listener can always perceive the sound image from the direction I gave "alSource3f(source, AL_POSITION, x,y,z)" , regardless of the speaker's position.（Assuming the speaker is always facing the listener.）

**Comment by kcat** [comment]
That sounds like something that would be best suited to applying on the final output, rather than per-source. OpenAL Soft doesn't really have anything for this, although you could probably make either a WASAPI APO for Windows or a PulseAudio module for Linux, which takes the left/right channel samples for a stereo device and applies the appropriate relative delays (and attenuation) given the listener's distance to each speaker.

Technically, OpenAL Soft does have per-speaker distance compensation on output, although it's not applied to stereo and it's not dynamically controllable (it depends on the ambisonic decoder definition's speaker distance properties, which gets applied when setting up the device and stays constant). Making it dynamic and exposing control of it to the app would need careful designing to ensure it can be set safely and stay compatible with future updates, considering OpenAL doesn't currently expose the output device configuration.

**Comment by Miyakawa-99** [comment]
Certainly, it is best to apply it in the final output.
However, I thought that the fastest way to achieve my goal is to use the echo effect.
This method may not be safe or accurate, but I think it can be done if I just add a left/right delay to the stereo sound.

So I tried this method, but the 3D sound I created and the echoes did not blend well.
If I add an echo, the pan given to the left and right will disappear.
I'll eventually implement the wasapiAPI you told me about, but if the method I'm trying to use isn't technically impossible, could you please tell me how to solve this problem?

**Comment by kcat** [comment]
It's not currently possible to make OpenAL Soft do it without changing the library. The best you could do is use OpenAL Soft's loopback extension to render a mix to your own memory buffer, apply the delays and attenuation to that buffer and output it using WASAPI or something manually.

**Comment by Miyakawa-99** [comment]
I understand.
I'm going to try more development based on your advice.
Thank you for your kindness.
